### PR TITLE
[V3 Fuzzy search] fix several issues with this feature

### PR DESCRIPTION
**Pull request by palmtree5**
### Type

- [x] Bugfix
- [x] Enhancement
- [ ] New feature

### Description of the changes
This PR fixes several issues with the fuzzy command search functionality released in beta 15
Namely:
- It can now be toggled (for the server by the server owner, and for DMs by the bot owner) (as requested in #1736)
- It now doesn't try to suggest anything when a valid custom command or alias is run
- It now checks that:
  - The user can run the suggested command and all parent commands of it
  - That the suggested command and all of its parents are not hidden
  - That the score for the match is >= 90
Non-alphanumeric characters were already being replaced with whitespace by the default processor, which was leading to results that had a score of 0. With the score requirement, these won't display anymore.

The one thing that still needs to be done is something about the `speedup` extra, which is technically optional but I feel should be used if at all possible. I still need to look into how I'd want to handle this part of it, so don't merge this quite yet

### CODE REVIEW

**Comment by palmtree5** [comment]
I believe I have figured out a way of detecting compiler availability, however this will need testing on all platforms (to the point of literally `pip install`ing this PR's branch).

**Comment by Eslyium** [comment]
Messed around with it for a good 15/20 minutes on a Ubuntu 16.04 server, I can say this is a whole lot better than what it was before in terms of searching and matching up with a command the user meant no issues so far.

**Comment by mikeshardmind** [comment]
This seems to meet the stated goals, and I have not found any issues with it currently

File: setup.py [hunk 1]
```diff
@@ -21,6 +23,28 @@ def get_package_list():
...
+def check_compiler_available():
+    env_dirs = os.environ["PATH"]
+    comp_name = ccompiler.get_default_compiler()
+    comp_info = ccompiler.compiler_class[comp_name]
+
+    mod_name = ".".join(["distutils", comp_info[0]])
+    mod = importlib.import_module(mod_name)
+
+    comp_class = getattr(mod, comp_info[1])
+
```
Comment (mikeshardmind) [inline setup.py]: This needs a check that the dict isn't empty

**Comment by palmtree5** [comment]
So we've now changed how we're handling compiler detection by using `tempfile` to create a file called `dummy.c` which is then compiled using `distutils`. If we can't compile that file for some reason, we'd drop `fuzzywuzzy`'s extra

**Review by Kowlin — APPROVED** [review]

**Pull request merged**